[Clang][AArch64][SVE] Allow write to SVE vector elements using the subscript operator

momchil-velikov opened the pull request:
The patch at https://reviews.llvm.org/D122732 introduced using the array subscript operator for SVE vectors, however it also causes an ICE when the subscripting expression is used as an lvalue.

This patches fixes the error. Lvalue subscripting expressions are emitted as LLVM IR `insertelement`.

efriedma-quic commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4180,8 +4180,10 @@ LValue CodeGenFunction::EmitArraySubscriptExpr(const ArraySubscriptExpr *E,
+  if (QualType BaseTy = E->getBase()->getType();
+      (BaseTy->isVectorType() && !isa<ExtVectorElementExpr>(E->getBase())) ||
+      (BaseTy->isBuiltinType() &&
+       BaseTy->getAs<BuiltinType>()->isSveVLSBuiltinType())) {
```
isSizelessVectorType()?

momchil-velikov replied:
AFAICT here https://github.com/llvm/llvm-project/blob/371eccd5dfed88c8e76449233d8388c12be3464b/clang/lib/Sema/SemaExpr.cpp#L5307 we are enabling array subscripts for SVE only. 

Perhaps we can be generalised to any size-less vector type, in a followup patch, @jacquesguan , what do you think?

efriedma-quic replied:
In that case, it might make sense to introduce a helper Type::isSubscriptableVectorType() helper to types, so we can easily find all the places that care about this.

momchil-velikov replied:
Split out to https://github.com/llvm/llvm-project/pull/92778

efriedma-quic commented on clang/test/CodeGen/aarch64-sve-vector-subscript-ops.c:
```diff
@@ -88,3 +88,13 @@ float subscript_float32(svfloat32_t a, size_t b) {
+// CHECK-NEXT:    ret <vscale x 4 x float> [[VECINS]]
+//
+svfloat32_t subscript_write_float32(svfloat32_t a, size_t b) {
+  a[b] = 1.0f;
```
Please also add a test for a compound op (`a[b]+=1.f`).

momchil-velikov replied:
Done.

DavidTruby reviewed (approved):
LGTM

momchil-velikov committed:
[Clang][AArch64][SVE] Allow write to SVE vector elements using the subscript operator

The patch at https://reviews.llvm.org/D122732 introduced using
the array subscript operator for SVE vectors, however it also
causes an ICE when the subscripting expression is used
as an lvalue.

This patches fixes the error. Lvalue subscripting expressions are
emitted as LLVM IR `insertvector`.

Change-Id: I46d0333d8ed8508cd9cd23e02dd1c2d48fb74cd2

momchil-velikov committed:
[fixiup] Add a test using compound assignment operator

Change-Id: I81e1fd4f23eb65a96e71015de7a4562fcbc53c0f

momchil-velikov commented:
Rebased.

efriedma-quic reviewed (approved):
LGTM
